sale_stock_mto_as_mts_orderpoint_bom/models/sale_order_line.py

```diff
@@ -0,0 +1,59 @@
+# Copyright 2024 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+
+from itertools import groupby
+
+from odoo import models
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+
+    def _run_orderpoints_for_mto_products(self):
+        orderpoints_to_procure_ids = []
+        sale_bom_lines = self.browse()
+
+        def all_moves_can_be_bought(moves):
+            for move in moves:
+                if not move.product_id.purchase_ok:
+                    return False
+            return True
+
+        def no_move_can_be_bought(moves):
+            for move in moves:
+                if move.product_id.purchase_ok:
+                    return False
+            return True
+
+        for line in self:
+            delivery_moves = line.move_ids.filtered(
+                lambda m: m.picking_id.picking_type_code == "outgoing"
+                and m.state not in ("done", "cancel")
+            )
+            sorted_moves = delivery_moves.sorted(lambda m: m.bom_line_id.bom_id)
+
+            if all_moves_can_be_bought(delivery_moves):
+                # if all moves are buyable, they should be bought individually
+                continue
+            for bom, move_list in groupby(sorted_moves, lambda m: m.bom_line_id.bom_id):
+                bom_moves = self.env["stock.move"].browse([m.id for m in move_list])
+                bom_lines = bom.bom_line_ids
+                if bom_lines != bom_moves.bom_line_id:
+                    continue
+                orderpoint = line._get_mto_orderpoint(bom.product_id)
+                if orderpoint.procure_recommended_qty:
+                    orderpoints_to_procure_ids.append(orderpoint.id)
+            sale_bom_lines |= line
+        wiz = (
+            self.env["make.procurement.orderpoint"]
+            .with_context(
+                **{
+                    "active_model": "stock.warehouse.orderpoint",
+                    "active_ids": orderpoints_to_procure_ids,
+                    "ignore_bom_find": True,
+                }
+            )
+            .create({})
+        )
+        wiz.make_procurement()
+        return super(SaleOrderLine, self - sale_bom_lines)._run_orderpoints_for_mto_products()
```

Comment:
```suggestion
# Copyright 2024 Camptocamp SA
# Copyright 2024 Michael Tietz (MT Software) <mtietz@mt-software.de>
# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
from odoo import models
from odoo.tools import groupby


class SaleOrderLine(models.Model):
    _inherit = "sale.order.line"

    def _run_orderpoints_for_mto_products(self):
        orderpoints_to_procure_ids = []
        sale_bom_lines = self.browse()

        def all_moves_can_be_bought(moves):
            for move in moves:
                if not move.product_id.purchase_ok:
                    return False
            return True

        def no_move_can_be_bought(moves):
            for move in moves:
                if move.product_id.purchase_ok:
                    return False
            return True

        for line in self:
            delivery_moves = line.move_ids.filtered(
                lambda m: m.picking_id.picking_type_code == "outgoing"
                and m.state not in ("done", "cancel")
                and m.is_from_mto_route
                and m.bom_line_id
            )
            if not delivery_moves:
                continue
            if all_moves_can_be_bought(delivery_moves):
                # if all moves are buyable, they should be bought individually
                continue
            for bom, move_list in groupby(
                delivery_moves, lambda m: m.bom_line_id.bom_id
            ):
                bom_moves = self.env["stock.move"].union(*move_list)
                bom_lines = bom.bom_line_ids
                if bom_lines != bom_moves.bom_line_id:
                    continue
                orderpoint = line._get_mto_orderpoint(bom.product_id)
                if orderpoint.procure_recommended_qty:
                    orderpoints_to_procure_ids.append(orderpoint.id)
            sale_bom_lines |= line
        if orderpoints_to_procure_ids:
            wiz = ( 
                self.env["make.procurement.orderpoint"]
                .with_context(
                    **{
                        "active_model": "stock.warehouse.orderpoint",
                        "active_ids": orderpoints_to_procure_ids,
                        "ignore_bom_find": True,
                    }
                )
                .create({})
            )
            wiz.make_procurement()
        return super(
            SaleOrderLine, self - sale_bom_lines
        )._run_orderpoints_for_mto_products()
```

Reply:
I updated the code, it was not working because also lines were used without any bom and which were not coming from mto.

Reply:
It's complicated to find the differences here....

Reply:
@mt-software-de I updated this method. Can you please redo this without selecting the whole block ? That'd be helpful.
Thanks !

Reply:
Also, I kind of disagree with this `_all_moves_can_be_purchased` method.
This has nothing to do on the sale.order.line model.
I could move that on stock.move if you really want me to extract that in a method.
But I'm quite ok with the inner function.

Reply:
yes i don't really care if it is inner or not. i just fixed the method in our testing
the main difference is the filter on delivery_moves, we don't want to do it for moves which do not have a bom_line_id set or not is_from_mto_route
```
        for line in self:
            delivery_moves = line.move_ids.filtered(
                lambda m: m.picking_id.picking_type_code == "outgoing"
                and m.state not in ("done", "cancel")
                and m.is_from_mto_route
                and m.bom_line_id
            )
            if not delivery_moves:
                continue
```
also i changed the groupby
instead of itertools the default from odoo.tools
this does not require a sorting
```
from odoo.tools import groupby
```